File: app/src/main/java/aeropresscipe/divinelink/aeropress/history/HistoryFragment.java
package aeropresscipe.divinelink.aeropress.history;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;

import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.List;

import aeropresscipe.divinelink.aeropress.R;
import aeropresscipe.divinelink.aeropress.generaterecipe.models.Recipe;
import aeropresscipe.divinelink.aeropress.timer.TimerActivity;
import aeropresscipe.divinelink.aeropress.timer.TimerFlow;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

public class HistoryFragment extends Fragment implements IHistoryView, ISharedPrefHistoryManager {

    private IHistoryPresenter presenter;

    private RecyclerView historyRecipesRV;
    private Toolbar mToolBar;
    private LinearLayout mEmptyListLL;
    private Animation mFadeAnimation;

    private boolean mHistoryIsEmpty;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View v = inflater.inflate(R.layout.fragment_history, container, false);

        historyRecipesRV = (RecyclerView) v.findViewById(R.id.historyRV);
        mToolBar = (Toolbar) v.findViewById(R.id.toolbar);
        mToolBar.setOnMenuItemClickListener(toolbarMenuClickListener);
        mEmptyListLL = (LinearLayout) v.findViewById(R.id.emptyListLayout);

        mToolBar.setNavigationOnClickListener(v1 -> getActivity().onBackPressed());


        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        historyRecipesRV.setLayoutManager(layoutManager);

        presenter = new HistoryPresenterImpl(this);
        presenter.getHistoryRecipes(getContext());

        return v;
    }

    public static HistoryFragment newInstance() {
        HistoryFragment fragment = new HistoryFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void showHistory(final List<History> savedRecipes) {


        if (getActivity() != null) {
            getActivity().runOnUiThread(new Runnable() {
                final HistoryRecipesRvAdapter historyRecipesRvAdapter = new HistoryRecipesRvAdapter(savedRecipes, getActivity(), historyRecipesRV);

                @Override
                public void run() {
                    mFadeAnimation = AnimationUtils.loadAnimation(getActivity(), R.anim.fade_in_favourites);
                    historyRecipesRV.setAnimation(mFadeAnimation);
                    historyRecipesRV.setAdapter(historyRecipesRvAdapter);
                    historyRecipesRvAdapter.createSwipeHelper();
                    historyRecipesRvAdapter.setPresenter(presenter);
                }
            });
        }
    }

    @Override
    public void showEmptyListMessage() {

        if (getActivity() != null) {
            getActivity().runOnUiThread(() -> {

                beginFading(historyRecipesRV, AnimationUtils.loadAnimation(getActivity(), R.anim.fade_out_favourites), 8);
                beginFading(mEmptyListLL, AnimationUtils.loadAnimation(getActivity(), R.anim.fade_in_favourites), 0);
                setIsHistoryEmptyBool(getContext(), true);

            });
        }
    }

    public void beginFading(View view, Animation animation, int visibility) {
        // 0 = View.VISIBLE, 4 = View.INVISIBLE, 8 = View.GONE
        view.startAnimation(animation);
        view.setVisibility(visibility);
    }

    @Override
    public void passData(Recipe recipe) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(() -> {
                recipe.setNewRecipe(true);
                startActivity(TimerActivity.newIntent(requireContext(), recipe, TimerFlow.START));
            });
        }
    }

    Toolbar.OnMenuItemClickListener toolbarMenuClickListener = new Toolbar.OnMenuItemClickListener() {
        @Override
        public boolean onMenuItemClick(MenuItem item) {
            switch (item.getItemId()) {
                case R.id.action_delete_history:

                    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getContext());
                    mHistoryIsEmpty = preferences.getBoolean("isHistoryEmpty", false);
                    //TODO Make button unavailable when History is empty.
                    if (!mHistoryIsEmpty)
                        showDeleteHistoryDialog();

                    return true;
                default:
                    return false;
            }
        }
    };

    public void showDeleteHistoryDialog() {
        new MaterialAlertDialogBuilder(getContext())
                .setTitle(R.string.action_delete)
                .setMessage(R.string.deleteHistoryMessage)
                .setPositiveButton(R.string.delete, (dialogInterface, i) -> presenter.clearHistory(getContext()))
                .setNegativeButton(R.string.cancel, (dialogInterface, i) -> {
                })
                .show();
    }

    @Override
    public void setIsHistoryEmptyBool(Context ctx, boolean bool) {
        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(ctx);
        final SharedPreferences.Editor editor = preferences.edit();
        editor.putBoolean("isHistoryEmpty", bool);
        editor.apply();
    }

    @Override
    public void setRecipeLiked(final boolean isLiked, final Integer pos) {

        if (getActivity() != null) {
            getActivity().runOnUiThread(() -> historyRecipesRV.getAdapter().notifyItemChanged(pos, isLiked));
        }

    }
}

File: app/src/main/java/aeropresscipe/divinelink/aeropress/history/HistoryFragment.kt
package aeropresscipe.divinelink.aeropress.history

import aeropresscipe.divinelink.aeropress.R
import aeropresscipe.divinelink.aeropress.base.HomeActivity.Companion.PAD_BOTTOM_OF_RECYCLER
import aeropresscipe.divinelink.aeropress.base.TimerViewCallback
import aeropresscipe.divinelink.aeropress.databinding.FragmentHistoryBinding
import aeropresscipe.divinelink.aeropress.recipe.models.Recipe
import aeropresscipe.divinelink.aeropress.savedrecipes.adapter.EmptyType
import aeropresscipe.divinelink.aeropress.timer.TimerFlow
import aeropresscipe.divinelink.aeropress.util.mapping.MappingAdapter
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.annotation.Px
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import dagger.hilt.android.AndroidEntryPoint
import gr.divinelink.core.util.extensions.padding
import gr.divinelink.core.util.extensions.setDisabled
import gr.divinelink.core.util.extensions.setEnabled
import gr.divinelink.core.util.swipe.ActionBindHelper
import gr.divinelink.core.util.swipe.SwipeAction
import gr.divinelink.core.util.utils.DimensionUnit
import java.lang.ref.WeakReference
import javax.inject.Inject

@AndroidEntryPoint
class HistoryFragment : Fragment(),
    HistoryStateHandler,
    IHistoryViewModel,
    TimerViewCallback {
    private var binding: FragmentHistoryBinding? = null

    private lateinit var callback: Callback

    @Inject
    lateinit var assistedFactory: HistoryViewModelAssistedFactory
    private lateinit var viewModel: HistoryViewModel

    private val historyAdapter = MappingAdapter()
    private var clearMenuItem: MenuItem? = null

    @Px
    private var recyclerViewPadding: Int = DimensionUnit.PIXELS.toPixels(PAD_BOTTOM_OF_RECYCLER).toInt()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentHistoryBinding.inflate(inflater, container, false)
        val view = binding?.root

        val viewModelFactory = HistoryViewModelFactory(assistedFactory, WeakReference<IHistoryViewModel>(this))
        viewModel = ViewModelProvider(this, viewModelFactory)[(HistoryViewModel::class.java)]
        viewModel.delegate = WeakReference(this)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding?.recyclerView?.padding(bottom = recyclerViewPadding)
        setToolbar()
        bindAdapter()
    }

    private fun setToolbar() {
        binding?.toolbar?.apply {
            setNavigationOnClickListener { activity?.onBackPressed() }
            inflateMenu(R.menu.history)
            setOnMenuItemClickListener { item ->
                clearMenuItem = binding?.toolbar?.menu?.findItem(R.menu.history)
                when (item.itemId) {
                    R.id.menu_clear -> {
                        viewModel.clearHistory(false)
                        true
                    }
                    else -> false
                }
            }
        }
    }

    private fun bindAdapter() {
        binding?.recyclerView?.apply {
            layoutManager = LinearLayoutManager(activity)
            adapter = historyAdapter
        }
        HistoryItem.register(historyAdapter,
            onActionClicked = { recipe: History, swipeAction: SwipeAction ->
                when (swipeAction.actionId) {
                    R.id.brew -> viewModel.startBrew(recipe.recipe)
                }
            },
            onLike = { recipe: History, position: Int ->
                viewModel.likeRecipe(recipe, position)
            },
            actionBindHelper = ActionBindHelper()
        )
    }

    override fun onResume() {
        super.onResume()
        viewModel.refresh()
    }

    override fun updateState(state: HistoryState) {
        when (state) {
            is HistoryState.ErrorState -> handleErrorState(state)
            is HistoryState.InitialState -> handleInitialState()
            is HistoryState.LoadingState -> handleLoadingState()
            is HistoryState.EmptyHistoryState -> handleEmptyHistoryState()
            is HistoryState.ShowHistoryState -> handleShowHistoryState(state)
            is HistoryState.StartNewBrewState -> handleStartNewBrewState(state)
            is HistoryState.RecipeLikedState -> handleRecipeLikedState(state)
            is HistoryState.ClearHistoryPopUpState -> handleClearHistoryPopUpState()
            is HistoryState.ShowSnackBar -> handleShowSnackBar(state)
        }
    }

    override fun handleInitialState() {
        // Intentionally Blank.
    }

    override fun handleLoadingState() {
        // Intentionally Blank.
    }

    override fun handleErrorState(state: HistoryState.ErrorState) {
        // Intentionally Blank.
    }

    override fun handleShowHistoryState(state: HistoryState.ShowHistoryState) {
        historyAdapter.submitList(state.list)
        updateToolbar(true)
    }

    override fun handleEmptyHistoryState() {
        historyAdapter.submitList(listOf(EmptyType.EmptyHistory))
        updateToolbar(false)
    }

    private fun updateToolbar(enable: Boolean) {
        clearMenuItem = binding?.toolbar?.menu?.findItem(R.id.menu_clear)
        when (enable) {
            true -> clearMenuItem?.setEnabled()
            false -> clearMenuItem?.setDisabled()
        }
    }

    override fun handleStartNewBrewState(state: HistoryState.StartNewBrewState) {
        callback.onUpdateRecipe(state.recipe, TimerFlow.START, true)
    }

    override fun handleRecipeLikedState(state: HistoryState.RecipeLikedState) {
        val items = historyAdapter.currentList.toMutableList()
        items[state.position] = state.item
        historyAdapter.submitList(items)
    }

    override fun handleClearHistoryPopUpState() {
        MaterialAlertDialogBuilder(requireContext())
            .setTitle(R.string.action_delete)
            .setMessage(R.string.deleteHistoryMessage)
            .setPositiveButton(R.string.delete) { _, _ ->
                viewModel.clearHistory(true)
            }
            .setNegativeButton(R.string.cancel) { _, _ ->
                // Intentionally Blank.
            }
            .show()
    }

    override fun handleShowSnackBar(state: HistoryState.ShowSnackBar) {
        callback.onSnackbarShow(state)
    }

    override fun updateBottomPadding(bottomPadding: Int) {
        recyclerViewPadding = bottomPadding
        binding?.recyclerView?.padding(bottom = bottomPadding)
    }

    companion object {
        @JvmStatic
        fun newInstance(): HistoryFragment {
            val fragment = HistoryFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        callback = context as Callback
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    interface Callback {
        fun onSnackbarShow(state: HistoryState.ShowSnackBar)
        fun onUpdateRecipe(recipe: Recipe, flow: TimerFlow, update: Boolean)
    }
}
